injector/injector.go

```diff
@@ -0,0 +1,312 @@
+					ti.provider.PopulateEventMacros(creativeId, "", "")
+
+					for _, url := range ti.events.VideoClicks {
+
+						outputXML.WriteString("<ClickTracking><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></ClickTracking>")
+					}
+
+					outputXML.WriteString("</VideoClicks>")
+
+				}
+				if injectTracker {
+					injectTracker = false
+					encoder.Flush()
+					outputXML.WriteString("<TrackingEvents>")
+					for typ, urls := range ti.events.LinearTrackingEvents {
+						ti.provider.PopulateEventMacros(creativeId, "", typ)
+						for _, url := range urls {
+							outputXML.WriteString("<Tracking event=\"")
+							outputXML.WriteString(typ)
+							outputXML.WriteString("\"><![CDATA[")
+							ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+							outputXML.WriteString("]]></Tracking>")
+						}
+					}
+					outputXML.WriteString("</TrackingEvents>")
+				}
+				encoder.EncodeToken(tt)
+				continue
+			case "InLine", "Wrapper":
+				wrapperTagFound = false
+				inlineWrapperTagFound = true
+				encoder.Flush()
+
+				if !impressionTagFound {
+					for _, url := range ti.events.Impressions {
+						outputXML.WriteString("<Impression><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></Impression>")
+					}
+				}
+				impressionTagFound = false
+
+				if !errorTagFound {
+					for _, url := range ti.events.Errors {
+						outputXML.WriteString("<Error><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></Error>")
+
+					}
+				}
+				errorTagFound = false
+
+				encoder.EncodeToken(tt)
+			case "NonLinear":
+				encoder.Flush()
+
+				ti.provider.PopulateEventMacros(creativeId, "", "")
+				for _, url := range ti.events.NonLinearClickTracking {
+					outputXML.WriteString("<NonLinearClickTracking><![CDATA[")
+					ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+					outputXML.WriteString("]]></NonLinearClickTracking>")
+				}
+				nonLineaerTagFound = true
+				encoder.EncodeToken(tt)
+			case "Companion":
+				companionTagFound = true
+				encoder.Flush()
+				ti.provider.PopulateEventMacros(creativeId, "", "")
+				for _, url := range ti.events.CompanionClickThrough {
+					outputXML.WriteString("<CompanionClickThrough><![CDATA[")
+					ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+					outputXML.WriteString("]]></CompanionClickThrough>")
+				}
+				encoder.EncodeToken(tt)
+			case "Creative":
+				isCreative = false
+			case "CompanionAds":
+				if !companionTagFound && wrapperTagFound {
+					outputXML.WriteString("<Companion>")
+					for _, url := range ti.events.CompanionClickThrough {
+						outputXML.WriteString("<CompanionClickThrough><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></CompanionClickThrough>")
+					}
+					outputXML.WriteString("<Companion>")
```

Comment:
Could you add tests to cover this block of code?

Reply:
Done